rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -220,180 +334,235 @@ void GraphCache::parse_put(const std::string & keyexpr)
-  add_topic_data(entity, *(node_it->second), *this);
+  // Then update the variables tracking topics across the graph.
+  if (entity.type() == EntityType::Publisher ||
+    entity.type() == EntityType::Subscription)
+  {
+    update_topic_map_for_del(this->graph_topics_, entity, true);
+  } else {
+    update_topic_map_for_del(this->graph_services_, entity);
+  }
 }
 
 ///=============================================================================
-void GraphCache::parse_del(const std::string & keyexpr)
+void GraphCache::update_topic_map_for_del(
+  GraphNode::TopicMap & topic_map,
+  const liveliness::Entity & entity,
+  bool report_events)
 {
-  std::optional<liveliness::Entity> valid_entity = liveliness::Entity::make(keyexpr);
-  if (!valid_entity.has_value()) {
-    // Error message has already been logged.
+  if (!entity.topic_info().has_value()) {
+    RCUTILS_LOG_WARN_NAMED(
+      "rmw_zenoh_cpp",
+      "update_topic_maps_for_del() called for non-node entity without valid TopicInfo. "
+      "Report this.");
     return;
   }
-  const liveliness::Entity & entity = *valid_entity;
-
-  // Helper lambda to update graph_topics_.
-  auto update_graph_topics =
-    [](const liveliness::TopicInfo topic_info, const EntityType entity_type, std::size_t pub_count,
-      std::size_t sub_count, GraphCache & graph_cache) -> void
-    {
-      GraphNode::TopicMap & graph_endpoints =
-        entity_type == EntityType::Publisher || entity_type == EntityType::Subscription ?
-        graph_cache.graph_topics_ :
-        graph_cache.graph_services_;
-      GraphNode::TopicMap::iterator cache_topic_it =
-        graph_endpoints.find(topic_info.name_);
-      if (cache_topic_it == graph_endpoints.end()) {
+  const liveliness::TopicInfo topic_info = entity.topic_info().value();
+  const bool is_pub = is_entity_pub(entity);
+  // Initialize a map that will be populated with any QoS events that may be detected.
+  std::unordered_map<liveliness::Entity, std::unordered_set<rmw_zenoh_event_type_t>>
+  local_entities_with_events = {};
+
+  GraphNode::TopicMap::iterator cache_topic_it =
+    topic_map.find(topic_info.name_);
+  if (cache_topic_it == topic_map.end()) {
+    // This should not happen.
+    RCUTILS_LOG_ERROR_NAMED(
+      "rmw_zenoh_cpp", "topic name %s not found in topic_map. Report this.",
+      topic_info.name_.c_str());
+    return;
+  } else {
+    GraphNode::TopicTypeMap::iterator cache_topic_type_it =
+      cache_topic_it->second.find(topic_info.type_);
+    if (cache_topic_type_it == cache_topic_it->second.end()) {
+      // This should not happen.
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_zenoh_cpp", "topic type %s not found in for topic %s. Report this.",
+        topic_info.type_.c_str(), topic_info.name_.c_str());
+      return;
+    } else {
+      const std::string qos_str = liveliness::qos_to_keyexpr(topic_info.qos_);
+      GraphNode::TopicQoSMap::iterator cache_topic_qos_it = cache_topic_type_it->second.find(
```

Comment:
I have the same basic thought about this function; it is long and pretty deeply nested.  Can we refactor it somehow?

Reply:
Done https://github.com/ros2/rmw_zenoh/pull/103/commits/47973be62df1ea92f5387f9a7c5d5d5caec1e227